### PR TITLE
Reduce stylesheet-related memory consumption and improve performance

### LINKED ISSUE #2366: 🍴 Huge memory consumption 🍴
### Current behavior:
image

image

image

image

### Desired behavior:

So for some time Im developing tests using Cypress.
I saw that sometimes I had huge lag in system.
I saw today in System Monitor, that Cypress is allocating new chunks of memory while not relasing old ones.
Im seeing this when running cypress show command: 
`CYPRESS_baseUrl=http://dev.xyz.com:3000 cypress open`
### Steps to reproduce:

1) Run Cypress with Electron 59
2) Let tests run and do their stuff

Acutally my test is clicking on dropdown with different currencies.
When it's clicked, page is relaoded and I check if current currency is correct.

```javascript
  it('checks list of currencies', () => {
    cy.fixture('currencyList.json')
      .then(currencyList => {
        for (let [currency, value] of currencyList) {
          cy.changeCurrency(currency)
          cy.get(':nth-child(4) > .nbar__menu-link')
            .then(element => {
              const elementValue = element.text().trim()
              expect(elementValue).to.equal(value)
            })

        }
      })
  })
```
### Versions

Cypress 3.1.0

## System info:
```bash
DISTRIB_ID=LinuxMint
DISTRIB_RELEASE=19
DISTRIB_CODENAME=tara
DISTRIB_DESCRIPTION="Linux Mint 19 Tara"
NAME="Linux Mint"
VERSION="19 (Tara)"
ID=linuxmint
ID_LIKE=ubuntu
PRETTY_NAME="Linux Mint 19"
VERSION_ID="19"
HOME_URL="https://www.linuxmint.com/"
SUPPORT_URL="https://forums.ubuntu.com/"
BUG_REPORT_URL="http://linuxmint-troubleshooting-guide.readthedocs.io/en/latest/"
PRIVACY_POLICY_URL="https://www.linuxmint.com/"
VERSION_CODENAME=tara
UBUNTU_CODENAME=bionic

```
## Hardware info:
```bash
H/W path       Device     Class          Description
====================================================
                          system         GL62 7QF (16J5.3)
/0                        bus            MS-16J5
/0/1                      memory         64KiB BIOS
/0/3e                     memory         16GiB System Memory
/0/3e/0                   memory         8GiB SODIMM DDR4 Synchronous 2133 MHz (0.5 ns)
/0/3e/1                   memory         8GiB SODIMM DDR4 Synchronous 2400 MHz (0.4 ns)
/0/42                     memory         256KiB L1 cache
/0/43                     memory         1MiB L2 cache
/0/44                     memory         6MiB L3 cache
/0/45                     processor      Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
/0/100                    bridge         Intel Corporation
/0/100/1                  bridge         Skylake PCIe Controller (x16)
/0/100/1/0                display        GM107M [GeForce GTX 960M]
/0/100/2                  display        Intel Corporation
/0/100/14                 bus            Sunrise Point-H USB 3.0 xHCI Controller
/0/100/14/0    usb1       bus            xHCI Host Controller
/0/100/14/0/4             input          USB Keyboard
/0/100/14/0/8             input          SteelSeries Rival 100 Gaming Mouse
/0/100/14/0/a             communication  Bluetooth wireless interface
/0/100/14/0/b             multimedia     USB 2.0 Webcam Device
/0/100/14/0/c             generic        USB2.0-CRW
/0/100/14/1    usb2       bus            xHCI Host Controller
/0/100/14.2               generic        Sunrise Point-H Thermal subsystem
/0/100/16                 communication  Sunrise Point-H CSME HECI #1
/0/100/17                 storage        Sunrise Point-H SATA Controller [AHCI mode]
/0/100/1c                 bridge         Sunrise Point-H PCI Express Root Port #1
/0/100/1c/0    wlp2s0     network        Intel Corporation
/0/100/1c.3               bridge         Sunrise Point-H PCI Express Root Port #4
/0/100/1c.3/0  enp3s0     network        QCA8171 Gigabit Ethernet
/0/100/1f                 bridge         Sunrise Point-H LPC Controller
/0/100/1f.2               memory         Memory controller
/0/100/1f.3               multimedia     Intel Corporation
/0/100/1f.4               bus            Sunrise Point-H SMBus
/0/0           scsi0      storage        
/0/0/0.0.0     /dev/sda   disk           275GB Crucial_CT275MX3
/0/0/0.0.0/1   /dev/sda1  volume         511MiB Windows FAT volume
/0/0/0.0.0/2   /dev/sda2  volume         255GiB EXT4 volume
/0/2           scsi3      storage        
/0/2/0.0.0     /dev/sdb   disk           1TB WDC WD10JPVX-22J
/0/2/0.0.0/1   /dev/sdb1  volume         233GiB EFI partition
/0/2/0.0.0/2   /dev/sdb2  volume         15MiB reserved partition
/0/2/0.0.0/3   /dev/sdb3  volume         150GiB Windows NTFS volume
/0/2/0.0.0/4   /dev/sdb4  volume         504MiB Windows NTFS volume
/0/2/0.0.0/5   /dev/sdb5  volume         454GiB Windows NTFS volume
/0/2/0.0.0/6   /dev/sdb6  volume         511MiB Windows FAT volume
/0/2/0.0.0/7   /dev/sdb7  volume         92GiB EXT4 volume
/1                        power          To Be Filled By O.E.M.

```

### UPDATE
I just checked it on Chrome browser.
Same thing is happening but much slower.

### PR DESCRIPTION
**Pull request by CoryDanielson**
closes #2366

### Issue
While debugging performance of cypress running against the application I am working on, I noticed the same memory issues mentioned in #2366 related to our stylesheets (multiple, thousands of lines each). I also noticed similar performance issues mentioned within the issue's conversation, especially related to `reduceText`, and specifically when accessing `rule.cssText`.

##### Performance bottleneck when copying stylesheets
Screen Shot 2019-04-29 at 5 12 46 PM

##### Memory leak related to multiple identical stylesheet strings
Screen Shot 2019-04-29 at 10 22 24 PM

### Changes
1. Memoized `getCssRulesString`
2. Memoized `makePathsAbsoluteToStylesheet`

After realizing that `getCssRulesString` and `makePathsAbsoluteToStylesheet` are idempotent (given the same inputs, they return the same outputs), it made sense to memoize them based on the the input stylesheet's `href`. In a worst case scenario, these changes would avoid thousands of calls to both of these methods, and replace them with lookups to a `WeakMap` ([_.memoize default Cache implementation](https://lodash.com/docs/4.17.11#memoize))

### Further optimizations
Based on these changes, it does feel like there's some opportunity for similar gains if a user has large stylesheets injected into the DOM via `<style>` tags. Because there's no `href` attribute in play here, a different solution would be required. This would eventually become a problem for apps that rely heavily on CSS-in-JS libraries that create `<style>` tags at runtime.

### CODE REVIEW

**Comment by CoryDanielson** [comment]
I've added a test to recreate the issue, and I've noticed an increase in performance, but not memory. I thought this might be a 2-for-1 fix, but it appears that the memory consumption occurs elsewhere related to snapshots.

There is a recommended set of changes suggested in #2366 that I'll try to pull in.

**Comment by CoryDanielson** [comment]
I've been looking at the memory for a bit and following the code, it seems that the memory leak starts here

https://github.com/cypress-io/cypress/blob/develop/packages/driver/src/cy/snapshots.coffee#L35-L47

This code is storing link snapshots as a "CSS Rules string", rather than just storing the link href. This causes `aut-iframe.js` to dump the styles onto the page via `<style>` tags rather than `<link>s`. I think that this probably makes snapshots load more quickly, but it introduces a memory leak in doing so.

https://github.com/cypress-io/cypress/blob/develop/packages/runner/src/iframe/aut-iframe.js#L112-L121

I updated this method to only return the `{ href }` locally, and it decreased memory usage by 50%

<details>
<summary>Before "Fix": Total memory usage is 105 mb</summary>
Screen Shot 2019-04-29 at 10 22 24 PM
</details>
<details>
<summary>After "Fix": Total memory usage is 52.6 mb</summary>
Screen Shot 2019-04-29 at 10 24 00 PM
</details>

**Comment by CoryDanielson** [comment]
I just realized that `makePathsAbsoluteToStylesheet` is also memoizable/idempotent (same inputs = same outputs). I memoized that method, and the memory usage is cut in half with no loss in functionality. I pushed those changes as well.

**Comment by brian-mann** [comment]
This is some good sleuthing. @chrisbreiding can you review as well?

My question is whether or not the functions are actually idempotent because I believe CSS rules can be changed programmatically with javascript at any point.

Even if this is the case, this could likely be detectable and we could then intercept or respond to the CSS changes to prevent needing to read the CSS rules on each snapshot.

**Also worth noting** this only affects running Cypress in interactive mode via `cypress open` since we don't take snapshots in `cypress run` mode.

**Comment by CoryDanielson** [comment]
Ahh, good point. CSS rules are not save from JS....
  - Would it be possible to watch for mutations of stylesheets and nuke the memoize cache in that case?
  - or force users with code that mutates stylesheets to opt out of memoization via an environment flag? This feels not great, but hopefully mutating CSS rules is as rare as I would hope it is 😰

**Comment by brian-mann** [comment]
I'm sure that we can either watch for mutations either with `MutationObserver` or by overriding the CSS rules setters (if they are writable). Haven't checked but entirely possible.

There are two methods: `insertRule` and `deleteRule` that we could likely tap into to invalidate the cache.

- https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet/insertRule
- https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet/deleteRule

**Comment by brian-mann** [comment]
We likely wouldn't have to invalidate the cache either... we could probably just store the CSS rules list using an array of arrays structure that corresponds to a string cache - and then update it incrementally as these methods are called. This is probably an over optimization but likely possible.

**Comment by CoryDanielson** [comment]
@brian-mann I updated the server to return a stylesheet with 2500 rules.

**Comment by CoryDanielson** [comment]
Changes similar to what was suggested by @Henrik-Buchholz in #2366 would avoid the concerns with modifying stylesheets in JS and preserve memory with negligible changes in performance.

**Comment by CoryDanielson** [comment]
How do ya'll feel about conditionally introducing this functionality via configuration/environment variable? Maybe adding like

> cacheExternalCss
> default false
> When set to true, external stylesheets `<link href="" />` will be cached by `href`, in order to improve performance. Only set this flag to true if you are certain that your JS code does not modify any external stylesheets at runtime.

If further optimizations result in these changes becoming depreciated, then the new config/environment variable could be dropped. By opting into these changes, there should be no impact to existing users' tests.

----

**Edit:** I looked into doing this for a bit, and it seems like adding a new environment variable as well as testing it would be a substantial amount of work. It seems like this would require some refactoring within `snapshots.coffee` that felt a bit out of my league as a random contributor. I hit a wall when attempting to run `4068_external_css_spec.js` with and without this `cacheExternalStylesheet` variable. It appears that setting the variable via `Cypress.env(k ,v)` in a `before` call is too late. I wasn't able to find a way to set environment variables for a specific test, either.

**Comment by CoryDanielson** [comment]
Awesome, thanks for the update. We're looking forward to use more Cypress for our testing. It's a fantastic tool

File: packages/driver/test/cypress/fixtures/issue-4068-external-css.html [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <link rel="stylesheet" href="/dynamically-server-generated/issue-4068.css" />
+  </head>
+  <body>
+    <h1 id="test">random font size test</h1>
+    <script>
+      // The .c# classname is tighty coupled to generated CSS file made in packages/driver/test/support/server.coffee
```
Comment (CoryDanielson) [inline packages/driver/test/cypress/fixtures/issue-4068-external-css.html]: typo here `tighty` should be `tightly`

File: packages/driver/test/support/server.coffee [hunk 2]
```diff
@@ -91,13 +91,12 @@ niv.install("react-dom@15.6.1")
     .status(500)
     .send("<html><body>server error</body></html>")
 
-  css2366 = ""
-  makeCss2366 = ->
-    css = _.range(1, 2501).map((n) -> ".c#{n} { font-size: #{Math.round(n / 10)}px; }").join("\n")
+  makeCss2366 = _.once ->
```
Comment (CoryDanielson) [inline packages/driver/test/support/server.coffee]: nice `_.once` here 👍

File: packages/driver/src/util/limited_map.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+class LimitedMap extends Map {
+  constructor (limit = 100) {
+    super()
+
+    this._limit = limit
+  }
+
+  set (key, value) {
+    if (this.size === this._limit) {
+      const firstKey = Array.from(this.keys())[0]
```
Comment (brian-mann) [inline packages/driver/src/util/limited_map.js]: No support in IE for `Array.from`. Safer to use the ol' low-dasherooskie

**Pull request merged**